File: app/assets/javascripts/components/basic_image_uploader_input.js
export default class BasicImageUploaderInput {
  constructor(elenemt) {
    this.$elenemt = $(elenemt)
    this.uploadAPIPath = this.$elenemt.data('upload-api-path')
    this.imageModelName = this.$elenemt.data('image-model-name')
    this.$uploadInput = this.$elenemt.find('[data-upload-input]')
    this.$imageThumbnail = this.$elenemt.find('.image-thumbnail')
    this.$imageIDInput = this.$elenemt.find('input.string')

    this.$uploadInput.fileupload({
      dataType: 'json',
      url: this.uploadAPIPath,
      paramName: `${this.imageModelName}[image]`,
      method: 'POST',
      dropZone: this.$input,
      maxNumberOfFiles: 1,
      progressall: (e, data) => {
      },
      done: (e, data) => {
        const id = data.result.id
        const thumbnailURL = data.result.thumbnail_url
        this.$imageThumbnail.find('img').attr('src', thumbnailURL)
        this.$imageIDInput.val(id)
      },
      error: () => {
      },
    })
  }
}

File: app/inputs/basic_image_uploader_input.rb
# frozen_string_literal: true
class BasicImageUploaderInput < SimpleForm::Inputs::Base
  include Rails.application.routes.url_helpers

  def input(_wrapper_options = nil)
    # rubocop:disable OutputSafety
    thumbnail_size = options[:thumbnail_size] || 'thumbnail'
    object.try("build_#{attribute_name}") if object.try(attribute_name).blank?
    image_model = object.try(attribute_name)
    image_model_name = image_model.class.name.underscore.tr('/', '_')

    template.content_tag :div, class: 'basic_image_uploader_input', data: {
      upload_api_path: try("api_#{image_model_name.pluralize}_path"),
      image_model_name: image_model_name,
      thumbnail_size: thumbnail_size
    } do

      template.concat <<-EOF.strip_heredoc.html_safe
        <div class="preview-area">
          <div class="image-thumbnail">
            <a href="#{image_model&.image&.url}" target="_blank">
              <img src="#{image_model&.image&.try(thumbnail_size)&.url}" />
            </a>
          </div>
        </div>
      EOF

      template.concat <<-EOF.strip_heredoc.html_safe
        <div class="upload-area">
          <div class="add-image">
            上傳圖片
            <input
              type=\"file\"
              accept=\"image/jpg,image/jpeg,image/gif,image/png\"
              form=""
              data-upload-input="ture"
              multiple
            >
          </div>
        </div>
      EOF

      template.concat @builder.hidden_field "belonging_#{attribute_name}_id", data: { uploader: :id_input }
    end
  end
end
